File: examples/python/comm-throttling/comm-throttling.py
# Copyright (c) 2010-2022. The SimGrid Team. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the license (GNU LGPL) which comes with this package.

from argparse import ArgumentParser
import sys

from simgrid import Engine, Actor, Comm, Mailbox, this_actor


def create_parser() -> ArgumentParser:
    parser = ArgumentParser()
    parser.add_argument(
        '--platform',
        type=str,
        required=True,
        help='path to the platform description'
    )
    return parser


def sender(mailbox: Mailbox):
    this_actor.info("Send at full bandwidth")

    # First send a 2.5e8 Bytes payload at full bandwidth (1.25e8 Bps)
    payload = Engine.clock
    mailbox.put(payload, int(2.5e8))

    this_actor.info("Throttle the bandwidth at the Comm level")
    # ... then send it again but throttle the Comm
    payload = Engine.clock
    # get a handler on the comm first
    comm: Comm = mailbox.put_init(payload, int(2.5e8))

    # let throttle the communication. It amounts to set the rate of the comm to half the nominal bandwidth of the link,
    # i.e., 1.25e8 / 2. This second communication will thus take approximately twice as long as the first one
    comm.set_rate(int(1.25e8 / 2)).wait()


def receiver(mailbox: Mailbox):
    # Receive the first payload sent at full bandwidth
    sender_time = mailbox.get()
    communication_time = Engine.clock - sender_time
    this_actor.info(f"Payload received (full bandwidth) in {communication_time} seconds")

    # ... Then receive the second payload sent with a throttled Comm
    sender_time = mailbox.get()
    communication_time = Engine.clock - sender_time
    this_actor.info(f"Payload received (throttled) in {communication_time} seconds")


def main():
    settings = create_parser().parse_known_args()[0]
    e = Engine(sys.argv)
    e.load_platform(settings.platform)

    mailbox = e.mailbox_by_name_or_create("Mailbox")

    Actor.create("sender", e.host_by_name("node-0.simgrid.org"), sender, mailbox)
    Actor.create("receiver", e.host_by_name("node-1.simgrid.org"), receiver, mailbox)

    e.run()


if __name__ == "__main__":
    main()

File: examples/java/comm_throttling/comm_throttling.java
/* Copyright (c) 2007-2025. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

import org.simgrid.s4u.*;

class sender extends Actor {
  Mailbox mailbox;
  public sender(Mailbox mailbox_) { mailbox = mailbox_; }
  public void run() throws SimgridException
  {
    Engine.info("Send at full bandwidth");

    /* - First send a 2.5e8 Bytes payload at full bandwidth (1.25e8 Bps) */
    Double payload = Engine.get_clock();
    mailbox.put(payload, 2.5e8);

    Engine.info("Throttle the bandwidth at the Comm level");
    /* - ... then send it again but throttle the Comm */
    payload = Engine.get_clock();
    /* get a handler on the comm first */
    Comm comm = mailbox.put_init(payload, 2.5e8);

    /* let throttle the communication. It amounts to set the rate of the comm to half the nominal bandwidth of the link,
     * i.e., 1.25e8 / 2. This second communication will thus take approximately twice as long as the first one*/
    comm.set_rate(1.25e8 / 2).await();
  }
}

class receiver extends Actor {
  Mailbox mailbox;
  public receiver(Mailbox mailbox) { this.mailbox = mailbox; }
  public void run() throws SimgridException
  {
    /* - Receive the first payload sent at full bandwidth */
    Double senderTime        = (Double)mailbox.get();
    double communicationTime = Engine.get_clock() - senderTime;
    Engine.info(String.format("Payload received (full bandwidth) in %f seconds", communicationTime));

    /*  - ... Then receive the second payload sent with a throttled Comm */
    senderTime        = (Double)mailbox.get();
    communicationTime = Engine.get_clock() - senderTime;
    Engine.info(String.format("Payload received (throttled) in %f seconds", communicationTime));
  }
}

public class comm_throttling {
  public static void main(String[] args)
  {
    Engine e = new Engine(args);
    e.load_platform(args[0]);

    Mailbox mbox = e.mailbox_by_name("Mailbox");

    e.host_by_name("node-0.simgrid.org").add_actor("sender", new sender(mbox));
    e.host_by_name("node-1.simgrid.org").add_actor("receiver", new receiver(mbox));

    e.run();

    // The following call is useless in your code, but our continuous integration uses it to track memleaks
    e.force_garbage_collection();
  }
}
